src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/Dependency/DependencyDag.cs

```diff
@@ -0,0 +1,797 @@
+            //
+            //     n                              n
+            //    / \                             |
+            //   d   |  will be simplified to     d
+            //    \  |                            |
+            //      d2                            d2
+            //
+            // This is because n needs d and d2, but d also needs d2, so
+            // fundamentally n only needs d.
+            //
+
+            var roots = allNodes.Where(x => x.Dependents.Count == 0).ToSet();
+            foreach (var node in roots)
+            {
+                SimplifyTransitiveDependencies(node);
+            }
+
+            return new DependencyDag(allNodes);
+        }
+
+        private static HashSet<Reference> GetAllDependencyReferences(
+            IDataExtensionRepository reps,
+            IDataExtensionReference der)
+        {
+            //
+            // return ALL dependencies of the given IDataExtensionReference,
+            // regardless of type.
+            //
+
+            var cycleErrors = der.DependencyState.Errors.Where(x => x.Code == ErrorCodes.EXTENSION_DependencyCycle);
+            if (cycleErrors.Any())
+            {
+                var cycleMessage = cycleErrors.First().ToString();
+                throw new InvalidOperationException(cycleMessage);
+            }
+
+            var deps = der.DependencyReferences;
+
+            //
+            // we are concatenating all of the dependencies found
+            // by the dependency state, as well as those explicitly
+            // declared on the references themselves.
+            //
+
+            return der.DependencyReferences.RequiredSourceDataCookerPaths
+                .Select(reps.GetSourceDataCookerReference)
+                .Cast<IDataExtensionReference>()
+                .Concat(
+                    deps.RequiredCompositeDataCookerPaths.Select(reps.GetCompositeDataCookerReference)
+                        .Cast<IDataExtensionReference>())
+                .Concat(
+                    deps.RequiredDataProcessorIds
+                        .Select(reps.GetDataProcessorReference)
+                        .Cast<IDataExtensionReference>())
+                .Concat(
+                    der.RequiredDataCookers
+                        .Select(
+                            x =>
+                            {
+                                var r = reps.TryGetDataCookerReference(x, out var c);
+                                Debug.Assert(r);
+                                return c;
+                            })
+                        .Cast<IDataExtensionReference>())
+                .Concat(
+                    der.RequiredDataProcessors
+                        .Select(reps.GetDataProcessorReference)
+                        .Cast<IDataExtensionReference>())
+                .Select(Reference.Create)
+                .ToSet();
+        }
+
+        private static void SimplifyTransitiveDependencies(Node node)
```

Comment:
>SimplifyTransitiveDependencies [](start = 28, length = 30)

I'm wondering if this would be more efficient as a recursive method. probably not a big deal, just seems like we're doing a bunch of extra work this way.

Reply:
oh, i didn't look closely enough at the Node structure. this might be fine. :)

---
In reply to: [591759896](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/54#discussion_r591759896) [](ancestors = 591759896)

Reply:
I don't know how big graphs can get, so I did it iteratively as I didn't want to potentially blow the stack.